[doc] Improved contributing PR guidelines (#22775)
xbmc/xbmc: docs/CONTRIBUTING.md

## Reviews
Reviews are a great way to get familiar with the codebase. If you would like to start with reviewing just use **[this handy github filter](https://github.com/xbmc/xbmc/pulls?q=is%3Apr+is%3Aopen+review%3Anone)** to see PRs that are in need of one (we have plenty!).

> ## Pull request making guidelines

Reviewer: ```suggestion
## Pull request submission guidelines
```
?

* **Create topic branches**. Never ever open a pull request from your master branch. **Ever!**
* **Code against master branch**. Unless release specific, all Kodi development occurs in the `master` branch. If it needs backporting it can be done after it hits master.
* **One topic branch per change**. Found something that needs fixing but is unrelated to the current work? Create a new topic branch and open another *pull request*.
* **Use a coherent commit history**. Make sure each individual commit in your *pull request* is meaningful and organized in logical chunks. Tidy up and squash commits before submitting.
* **Be descriptive, but concise**. The *pull request* and *commits* should have an useful title and follow the format:

> `[Component(s)] Short description...` or alternative format `Component(s): Short description...`<br/>

Reviewer: One important thing not many people do is to keep a concise description on the commit message (not the title but the message below). That's pretty much the only way the context is kept in git history without being bound to a specific git platform like github (unlike the PR body, issues, git comments, etc). I would say that if the contribution contains code that need additional explanation (why that change is being done and what others should take into consideration if ever considering changing that piece of code) a proper commit message should be added as well.
I don't see the need to make it mandatory for all commits (only on particular circumstances). Otherwise if we ever move from github to a different platform there is no guarantee that all the other context is kept.

Author: modified take a second look if you prefer reword in other way

> by component we mean for example, the name of the modified class, the code section context (e.g. name of the parent folder), dependency name, etc... You can get hints by browsing the git history of the modified file, an example:<br/>

Reviewer: I think it is more important to identify the area of the code (gui, video, depends, build, etc) than the modified class itself. That can be easily found in git history

Author: modified take a second look

`[Python][xbmcgui] Fix x problem...` or `Python/xbmcgui: Fix x problem...`<br/>

> instead something like `[README.md] Update` is **not** descriptive enough.<br/>

Reviewer: ```suggestion
instead of something like `[README.md] Update` is **not** descriptive enough.<br/>
```

The *pull request* description should only contain information relevant to the change.
* **Draft PR**. If your work requires further development, and you wish to ask the opinion of other developers on how to proceed, you can publish your PR as Draft.

## Pull request merge guidelines

> * **First rule, be patient, be kind, rewind**. We're all volunteers with busy lives but usually very responsive in cataloging *pull requests* and pinging the relevant maintainer(s).

Reviewer: Probably we can refer to code owners here. If you touch specific areas/folders of the project automatic reviews are request to the appropriate code owners. Only if none is available we ping the relevant people that have some knowledge about the affected parts. This is also important as we need more code owners, external contributors shouldn't need to be aware of who maintains what what parts of the code are changed :)

Author: can you provide an example to add this point?

Reviewer: Although not valid for the whole codebase, specific components of Kodi have a set of [maintainers/code owners](https://github.com/xbmc/xbmc/blob/master/.github/CODEOWNERS) whose review will be automatically requested when you open a pull request.

Author: added

> * **Jenkins code formatting requests**. Always apply code style formatting requests from Jenkins. If not, please provide a valid reason why you do not apply them.

Reviewer: We should probably consider to provide a tutorial on how to integrate that with git so that they are automatically applied before pushing. Not asking you to do it :) It's more in the context of our git tutorial rather than here

Author: yeah not detailed example here or become heavy to read

> * **Jenkins CI builds**. It is mandatory to have green light on all Jenkins building platforms.

Reviewer: I think we can also mentions the jenkins labels here: Set "No-Jenkins" if your pull request is yet not complete or you don't want to have jenkins building the PR code.

Author: added but not team members can add this flag?

Reviewer 2: They can. What non-teammates can not do is to trigger builds with "Jenkins build this please"

Reviewer: No, only team members can add them. Regardless, it can be written in some sort of "agnostic" way: "No-Jenkins label can be set bla bla bla"

Reviewer: ok probably they can :)

Author: fine changed to your suggest change

* **Reviewing time**. Be patient, in general terms we speak of weeks not days, the active team is small and that means things can stack up, but  we are all doing our best. If your PR has been standing for too long without a review feel free to let us know.

> * **Reviewing process**. At least one member's approval is required to have green light to merge, to take in account that expertise is spread thinly across areas of Kodi, this makes getting reviews of some changes even harder. Kodi is tangled and what looks perfect to you may have unwanted side effects, so the approval by several members sometime is required, but in any case is always helpful, in particular for huge additions/refactors that may cause regressions. Therefore it is possible that your code will require more love, so more changes before its final version may be required, so arm yourself with relaxing herbal teas.

Reviewer: ```suggestion
* **Reviewing process**. At least one member's approval is required to have green light to merge, to take into account (...)
```

by component we mean for example, identify the area of the code (gui, video, depends, builds, etc), the code context (e.g. class name), etc... you can easily find hints by browsing through the git history, an example:<br/>
`[Python][xbmcgui] Fix x problem...` or `Python/xbmcgui: Fix x problem...`<br/>
instead of something like `[README.md] Update` is **not** descriptive enough.<br/>
Also adding a short description to the *commit* (under the title) is a good thing to consider to give a better explanation of the commit changes.<br/>
The *pull request* description should only contain information relevant to the change.
* **Draft PR**. If your work requires further development, and you wish to ask the opinion of other developers on how to proceed, you can publish your PR as Draft.

## Pull request merge guidelines
* **First rule, be patient, be kind, rewind**. We're all volunteers with busy lives but usually very responsive in cataloging *pull requests* and pinging the relevant maintainer(s).
* **Jenkins code formatting requests**. Always apply code style formatting requests from Jenkins. If not, please provide a valid reason why you do not apply them.

> * **Jenkins CI builds**. It is mandatory to have green light on all Jenkins building platforms. Set "No-Jenkins" flag if your pull request is yet not complete or you do not want to have jenkins building the PR code.

Reviewer: ```suggestion
* **Jenkins CI builds**. It is mandatory to have green light on all Jenkins building platforms. The "No-Jenkins" label can be set by maintainers to avoid having the CI building the PR code (and wasting build resources) in cases the contribution is still work in progress or if known in advance it won't build.
```

## Reviews
Reviews are a great way to get familiar with the codebase. If you would like to start with reviewing just use **[this handy github filter](https://github.com/xbmc/xbmc/pulls?q=is%3Apr+is%3Aopen+review%3Anone)** to see PRs that are in need of one (we have plenty!).

## Pull request submission guidelines
* **Create topic branches**. Never ever open a pull request from your master branch. **Ever!**
* **Code against master branch**. Unless release specific, all Kodi development occurs in the `master` branch. If it needs backporting it can be done after it hits master.
* **One topic branch per change**. Found something that needs fixing but is unrelated to the current work? Create a new topic branch and open another *pull request*.
* **Use a coherent commit history**. Make sure each individual commit in your *pull request* is meaningful and organized in logical chunks. Tidy up and squash commits before submitting.

> * **Be descriptive, but concise**. The *pull request* and *commits* should have an useful title and follow the format:<br/>

Reviewer 3: ```suggestion
* **Be descriptive, but concise**. The *pull request* and *commits* should have a useful title and follow the format:<br/>
```
"a useful title" is correct.

Author: I was thinking "an" was used before each vowel i remember wrong

Reviewer 3: Yes, that is the general rule, but it applies to the sound and not the writing. "Useful" is pronounced /ˈjuːs.fəl/ and starts with the consonant "j", which is why it is "a useful [something]" :-)

See also https://dictionary.cambridge.org/grammar/british-grammar/a-an-and-the

`[Component(s)] Short description...` or alternative format `Component(s): Short description...`<br/>

> by component we mean for example, identify the area of the code (gui, video, depends, builds, etc), the code context (e.g. class name), etc... you can easily find hints by browsing through the git history, an example:<br/>

Reviewer 3: ```suggestion
By "component", we mean that you should identify the area of the code (gui, video, depends, builds, etc), the code context (e.g. class name), etc. You can easily find hints by browsing through the git history. An example:<br/>
```

`[Python][xbmcgui] Fix x problem...` or `Python/xbmcgui: Fix x problem...`<br/>

> instead of something like `[README.md] Update` is **not** descriptive enough.<br/>

Reviewer 3: ```suggestion
instead of something like `[README.md] Update`, which  is **not** descriptive enough.<br/>
```

> Also adding a short description to the *commit* (under the title) is a good thing to consider to give a better explanation of the commit changes.<br/>

Reviewer 3: ```suggestion
Also, adding a short description to the *commit* (under the title) is a good thing to consider to give a better explanation of the commit changes.<br/>
```

The *pull request* description should only contain information relevant to the change.
* **Draft PR**. If your work requires further development, and you wish to ask the opinion of other developers on how to proceed, you can publish your PR as Draft.

## Pull request merge guidelines
* **First rule, be patient, be kind, rewind**. We're all volunteers with busy lives but usually very responsive in cataloging *pull requests* and pinging the relevant maintainer(s). Although not valid for the whole codebase, specific components of Kodi have a set of [maintainers/code owners](https://github.com/xbmc/xbmc/blob/master/.github/CODEOWNERS) whose review will be automatically requested when you open a pull request.
* **Jenkins code formatting requests**. Always apply code style formatting requests from Jenkins. If not, please provide a valid reason why you do not apply them.

> * **Jenkins CI builds**. It is mandatory to have green light on all Jenkins building platforms. The "No-Jenkins" label can be set by maintainers to avoid having the CI building the PR code (and wasting build resources) in cases the contribution is still work in progress or if known in advance it won't build.

Author: changed with green status

> * **Reviewing time**. Be patient, in general terms we speak of weeks not days, the active team is small and that means things can stack up, but  we are all doing our best. If your PR has been standing for too long without a review feel free to let us know.

Reviewer 3: ```suggestion
* **Reviewing time**. Be patient, in general terms we speak of weeks not days, the active team is small and that means things can stack up, even as we are all doing our best. If your PR has been standing for too long without a review, feel free to let us know.
```

> * **Reviewing process**. At least one member's approval is required to have green light to merge, to take into account that expertise is spread thinly across areas of Kodi, this makes getting reviews of some changes even harder. Kodi is tangled and what looks perfect to you may have unwanted side effects, so the approval by several members sometime is required, but in any case is always helpful, in particular for huge additions/refactors that may cause regressions. Therefore it is possible that your code will require more love, so more changes before its final version may be required, so arm yourself with relaxing herbal teas.

Reviewer 3: ```suggestion
* **Reviewing process**. At least one member's approval is required to have the green light to merge. Please take into account that expertise is spread thinly across areas of Kodi. This makes getting reviews of some changes even harder. Kodi is tangled and what looks perfect to you may have unwanted side effects, so the approval by several members sometime is required. In any case, this is always helpful, in particular for huge additions/refactors that may cause regressions. Therefore, it is possible that your code will require more love, i.e., more changes before its final version may be required, so arm yourself with relaxing herbal teas.
```

> * **Test builds**. When reasonable providing test builds to test your changes can speed up the review process, especially if you consider that some users or team members cannot build Kodi for testing. If you are a member of the kodi team, we remind you that there are jenkins helpers that can build and upload onto the mirrors where users can access them to test it. If you are not part of the team, you can make a request for test builds.

Reviewer 3: ```suggestion
* **Test builds**. When reasonable, providing test builds to test your changes can speed up the review process, especially if you consider that some users or team members cannot build Kodi for testing. If you are a member of Team Kodi, we remind you that there are Jenkins helpers that can build and upload onto the mirrors where users can access them to test it. If you are not part of the team, you can make a request for test builds.
```

> * **Test builds**. When reasonable, providing test builds to test your changes can speed up the review process, especially if you consider that some users or team members cannot build Kodi for testing. If you are a member of Team Kodi, we remind you that there are Jenkins helpers that can build and upload onto the mirrors where users can access them to test it. If you are not part of the team, you can make a request for test builds.

Reviewer 4: ... that can build and upload code onto the mirrors where users can access it for test purposes.

## Coding guidelines
* **Follow our [code guidelines](CODE_GUIDELINES.md)**. New code should follow those guidelines even if existing code doesn't. If you're up to it, improve existing code and submit it in separate commits.

> * **Document the code**. When changes are made to the code, its functionality is clear to the author, but unlike other users who read the block of code (even after several years) may not be facilitated. Providing documentation saves everyone the time spent studying a block of code over time.

Reviewer 3: ```suggestion
* **Document the code**. When changes are made to the code, its functionality may be clear to the author, but other developers who read the block of code (even after several years) can easily have a hard time understanding it. Providing documentation saves time for everyone.
```

as a side-note, after several years I'm quite sure that even being the original author of the code will not help you understand it any more :-)
